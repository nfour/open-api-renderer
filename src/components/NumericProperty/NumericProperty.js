import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

export default class NumericProperty extends PureComponent {
  render() {
    const { constraints } = this.props;

    if (!constraints) {
      return null;
    }

    const { exclusiveMinimum, exclusiveMaximum, maximum, minimum, multipleOf } = constraints;
    const validations = [];

    if (multipleOf) {
      validations.push(`multiple of ${multipleOf}`);
    }

    // We're following JSON-Schema Draft 6, which states that exclusive* are
    // integers, not boolean values. Also using `undefined` to prevent edge
    // cases where the value is 0 or 1.
    if (maximum !== undefined && minimum !== undefined) {
      validations.push(`${minimum}…${maximum}`);
    } else if (exclusiveMaximum !== undefined && exclusiveMinimum !== undefined) {
      validations.push(`${exclusiveMinimum}…${exclusiveMaximum}`);
    } else if (minimum !== undefined) {
      validations.push(`≥ ${minimum}`);
    } else if (maximum !== undefined) {
      validations.push(`≤ ${maximum}`);
    } else if (exclusiveMinimum !== undefined) {
      validations.push(`> ${exclusiveMinimum}`);
    } else if (exclusiveMaximum !== undefined) {
      validations.push(`< ${exclusiveMaximum}`);
    }

    if (!validations.length) {
      return null;
    }

    return (
      <span>
        {validations.map(constraint =>
          <span key={constraint} className="numeric-constraints">{constraint}</span>
        )}
      </span>
    );
  }
}

NumericProperty.propTypes = {
  constraints: PropTypes.shape({
    exclusiveMinimum: PropTypes.number,
    exclusiveMaximum: PropTypes.number,
    maximum: PropTypes.number,
    minimum: PropTypes.number,
    multipleOf: PropTypes.number
  })
};
