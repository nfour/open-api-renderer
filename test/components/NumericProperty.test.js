import React from 'react';
import NumericProperty from './../../src/components/NumericProperty/NumericProperty';
import renderer from 'react-test-renderer';

describe('<NumericProperty />', () => {
  it('renders nothing if there are no applicable constraints', () => {
    let tree = renderer.create(
      <NumericProperty />
    ).toJSON();

    expect(tree).toBeNull();

    tree = renderer.create(
      <NumericProperty constraints={{minItems: 1}}/>
    ).toJSON();

    expect(tree).toBeNull();
  });

  it('renders multipleOf', () => {
    const tree = renderer.create(
      <NumericProperty constraints={{multipleOf: 2}} />
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders maximum', () => {
    const tree = renderer.create(
      <NumericProperty constraints={{maximum: 2}} />
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders minimum', () => {
    const tree = renderer.create(
      <NumericProperty constraints={{minimum: 2}} />
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders minimum and maximum', () => {
    const tree = renderer.create(
      <NumericProperty constraints={{minimum: 2, maximum: 5}} />
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders exclusiveMaximum', () => {
    const tree = renderer.create(
      <NumericProperty constraints={{exclusiveMaximum: 2}} />
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders exclusiveMinimum', () => {
    const tree = renderer.create(
      <NumericProperty constraints={{exclusiveMinimum: 2}} />
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders exclusiveMinimum and exclusiveMaximum', () => {
    const tree = renderer.create(
      <NumericProperty constraints={{exclusiveMinimum: 2, exclusiveMaximum: 5}} />
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders a combination of constraints', () => {
    const tree = renderer.create(
      <NumericProperty constraints={{multipleOf: 2, minimum: 2, maximum: 20 }} />
    );

    expect(tree).toMatchSnapshot();
  });
});
