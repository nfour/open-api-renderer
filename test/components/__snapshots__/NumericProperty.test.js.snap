// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NumericProperty /> renders a combination of constraints 1`] = `
<span>
  <span
    className="numeric-constraints"
  >
    multiple of 2
  </span>
  <span
    className="numeric-constraints"
  >
    2…20
  </span>
</span>
`;

exports[`<NumericProperty /> renders exclusiveMaximum 1`] = `
<span>
  <span
    className="numeric-constraints"
  >
    &lt; 2
  </span>
</span>
`;

exports[`<NumericProperty /> renders exclusiveMinimum 1`] = `
<span>
  <span
    className="numeric-constraints"
  >
    &gt; 2
  </span>
</span>
`;

exports[`<NumericProperty /> renders exclusiveMinimum and exclusiveMaximum 1`] = `
<span>
  <span
    className="numeric-constraints"
  >
    2…5
  </span>
</span>
`;

exports[`<NumericProperty /> renders maximum 1`] = `
<span>
  <span
    className="numeric-constraints"
  >
    ≤ 2
  </span>
</span>
`;

exports[`<NumericProperty /> renders minimum 1`] = `
<span>
  <span
    className="numeric-constraints"
  >
    ≥ 2
  </span>
</span>
`;

exports[`<NumericProperty /> renders minimum and maximum 1`] = `
<span>
  <span
    className="numeric-constraints"
  >
    2…5
  </span>
</span>
`;

exports[`<NumericProperty /> renders multipleOf 1`] = `
<span>
  <span
    className="numeric-constraints"
  >
    multiple of 2
  </span>
</span>
`;
